### ISSUE TITLE
optimizing perfetto to process large trace files

### ISSUE THREAD

**Issue by Philius** [google/perfetto#539]
Here's the line
https://github.com/google/perfetto/blob/207bb3eb405653a9bd53fe9914fd25d2384af989/src/trace_processor/read_trace_internal.cc#L97

It's already doing PROT_READ, not sure what the "read_only" checkbox does.

The trace file `chrome-clang.json` is 32.2GB, too large for a private `mmap()` on my system.
`trace_processor` reported `Cannot use mmap on this system. Falling back on read()`

Note that I had previously opened an old version of the trace file chrome-clang.json that was 187MB in size in another tab, but I'm sure that's a different problem, I didn't know about this limitation and also didn't know it was open in another tab, which may have caused the crash.
```
cd Debug && /v3c/notes/clang/trace_processor --httpd /v3c/notes/chrome-clang.json
[415.834] d_trace_internal.cc:113 Cannot use mmap on this system. Falling back on read()
[415.834] processor_shell.cc:1641 Trace loaded: 0.00 MB in 0.00s (0.0 MB/s)
[415.835]             httpd.cc:99 [HTTP] Starting RPC server on localhost:9001
[415.835]            httpd.cc:104 [HTTP] This server can be used by reloading https://ui.perfetto.dev and clicking on YES on the "Trace Processor native acceleration" dialog or through the Python API (see https://perfetto.dev/docs/analysis/trace-processor#python-api).
[589.446]       http_server.cc:83 [HTTP] New connection
[589.446]      http_server.cc:231 [HTTP] POST /status [body=0B, origin="https://ui.perfetto.dev"]
[601.560]      http_server.cc:231 [HTTP] POST /status [body=0B, origin="https://ui.perfetto.dev"]
[601.564]       http_server.cc:83 [HTTP] New connection
[601.564]      http_server.cc:231 [HTTP] GET /websocket [body=0B, origin="https://ui.perfetto.dev"]
[601.565] e_processor_impl.cc:708 Resetting DB to initial state, deleting table/views:
[771.377]       http_server.cc:90 [HTTP] Client disconnected
[787.152]      http_server.cc:231 [HTTP] POST /status [body=0B, origin="https://ui.perfetto.dev"]
[822.870]      http_server.cc:231 [HTTP] POST /status [body=0B, origin="https://ui.perfetto.dev"]
[822.880]       http_server.cc:83 [HTTP] New connection
[822.880]      http_server.cc:231 [HTTP] GET /websocket [body=0B, origin="https://ui.perfetto.dev"]
Segmentation fault
```
I guess the 187MB was cached in html5 local storage in the other tab?

Anyway it then proceeded to carry on with the 187MB file, not the desired behaviour.

**Comment by LalitMaganti** [comment]
It's not clear to me from the bug what the problem actually is here. The title of the bug mentions "MAP_SHARED" but then the actual description does not really mention that at all. The "Cannot use mmap on this system. Falling back on read()" message is a warning more than an error so it's not really a problem if that fails.

**Comment by primiano** [comment]
I am not sure that MAP_PRIVATE vs MAP_SHARED will make any diffrence whatsoever.
The main difference between the two is:
- If this was a writable mapping, whether writes are propagated into the files (MAP_SHARED) or create copy-on-write anonymous memory overlays. But this doesn't apply as the mapping is read-only, as per PROT_READ.
- If the process fork()s, whether the written pages are propagated (MAP_SHARED) or not (MAP_PRIVATE).

In the case of a read-only mapping, to the best of my knowledge, shared vs private shouldn't make any difference whatsoever. 

This bug as is is very unactionable. what would make it more actionable is:

1) Details about your system. Which cpu? is it a 64 or 32 bit machine? On a 64-bit machine there's virtually no limit on how big of a file you can mmap, as that takes only virtual address space, not physical. On a 32-bit machine there's nothing we realistically can do (or any option is unnecessarily complex)

2) If you replace MAP_PRIVATE with MAP_SHARED does your problem go away? (I'd be extremely surprised if it did)

**Comment by Philius** [comment]
Good points.
I'm running Debian bookworm amd64.
```
$ cat /proc/meminfo
MemTotal:       16240124 kB
MemFree:         9158548 kB
...
SwapTotal:      33554428 kB
SwapFree:       33554428 kB
```

I just tried to reproduce the problem with a test program.
Here's mmap-json.cpp
```
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <unistd.h>
#include <iostream>
using namespace std;

int main(int argc, char * argv[])
{
  long page_size = sysconf(_SC_PAGESIZE);
  if(page_size != 4096) {
    cerr << "sysconf(_SC_PAGESIZE) = " << page_size << " != 4096" << endl;
    return 1;
  }
  int map_type = (argc == 1) ? MAP_PRIVATE : MAP_SHARED;
  cout << "map_type == " << (map_type == MAP_PRIVATE ? "MAP_PRIVATE" : "MAP_SHARED") << '.' << endl;
  int fd = open("chrome-clang.json", O_RDONLY);
  if(fd == -1) {
    cerr << "File could not be opened!" << endl;
    return 1;
  }
  size_t sz = lseek(fd, 0, SEEK_END), processed = 0;
  lseek(fd, 0, SEEK_SET);
  struct Page { uint8_t bytes[4096]; };
  Page * file_mm0 = (Page *)mmap(nullptr, sz, PROT_READ, map_type, fd, 0);
  unsigned int check = 0;
  if (file_mm0 != MAP_FAILED) {
    Page * file_mm = file_mm0;
    Page page;
    while((file_mm - file_mm0) < (sz/sizeof(Page))) {
      page = *file_mm++;
      check += page.bytes[0];
      ++processed;
    }
  }
  if(!processed) {
    cerr << "No pages processed !?!" << endl;
    return 1;
  }
  cout << "Processed " << processed << " pages, check = " << check << '.' << endl;
  munmap(file_mm0, sz);
  close(fd);
  return 0;
}
```
Both map_types work !?!
Somehow I got it into my head that MAP_PRIVATE is limited by physical + swap memory, while MAP_SHARED isn't.
But as you can see, even that's not a problem on my system !?!
```
$ ./mmap-json
map_type == MAP_PRIVATE.
Processed 8440987 pages, check = 628461615.
```
```
$ ./mmap-json 1
map_type == MAP_SHARED.
Processed 8440987 pages, check = 628461615.
```
So how did `trace_processor` report `Cannot use mmap on this system. Falling back on read()` ???

**Comment by primiano** [comment]
> Somehow I got it into my head that MAP_PRIVATE is limited by physical + swap memory, while MAP_SHARED isn't.

Hmm pagefaults are handled lazily though, memory is allocated only when trying to wirte (Which wouldn't work here).
You can check the state of /proc/sys/vm/overcommit_memory  but in my experience all modern systems have overcommit enable, which means the kernel will allow you vm-allocate way more memory than what physically available, and eventually invoke lowmemkiller and other vmscanners to get pages on-demand when running out (or kill the process)

I am confused by your repro though.
Are you saying that the same file gets mmap-ed correctly by your test program, but fails in trace processor?

Something else in that branch might fail then, mind attaching a debugger, or adding some printfs and checking?

I've never seen this problem before.

Maybe before doing that try playing around with /proc/sys/vm/overcommit_memory

**Comment by Philius** [comment]
I found the cached version of the old trace file in `Storage -> Cached Storage -> /cached_traces/...`.
Where's the executable cached?

**Comment by Philius** [comment]
Got heem!
`cd Debug && /v3c/notes/clang/trace_processor --httpd /v3c/notes/chrome-clang.json`
should have been
`cd Debug && /v3c/notes/clang/trace_processor --httpd /v3c/notes/clang/chrome-clang.json`
My system is currently chomping on swap.
Did not finding the file cause ui.perfetto.dev to use the cached version?

**Comment by Philius** [comment]
Got heem!
`cd Debug && /v3c/notes/clang/trace_processor --httpd /v3c/notes/chrome-clang.json`
should have been
`cd Debug && /v3c/notes/clang/trace_processor --httpd /v3c/notes/clang/chrome-clang.json`
https://github.com/google/perfetto/blob/207bb3eb405653a9bd53fe9914fd25d2384af989/src/trace_processor/read_trace_internal.cc#L82
should be changed to
```
if(fd == -1)
```

**Comment by primiano** [comment]
> should be changed to if(fd == -1)

I really doubt so. `fd` there is a ScopedFd not an Int. The bool operator of ScopedFd checks for != -1.

TBH I don't understand anymore what's your problem and what you are trying to do.
But I'm glad we closed this issue.

**Comment by Philius** [comment]
I accidentally gave it the name of a file that doesn't exist and it said it couldn't mmap() it.
And you're saying that's not a problem, and that I'm not making sense.
Bravo!
Have you tried giving it the name of a non-existing file and seeing if it can mmap() the file?
Because then you would be reproducing the problem and your confidence would shoot up from "pretty sure" to "certain".
To be fair, I didn't know about the operator bool(), but somewhere it's certainly not detecting that the file doesn't exist, which, I'm sure you'd agree, is a lot different to "can't mmap() the file".

**Comment by LalitMaganti** [comment]
Trace processor already checks if the file exists: you can see that check right [here](https://github.com/google/perfetto/blob/207bb3eb405653a9bd53fe9914fd25d2384af989/src/trace_processor/read_trace_internal.cc#L83).

If I pass a path to a file which does not exist, I get the message:
`Could not read trace file (path: /tmp/foo): Could not open trace file (path: /tmp/foo)`

**Comment by primiano** [comment]
Are you sure you are not passing a file mounted on a USB stick or network filesystem?

Thinking more I think you can stumble in that error if we are able to open() the file, but mmap fails.
mmap will fail over network file systems.

**Comment by Philius** [comment]
So you checked that opening a file that doesn't exist fails.

**Comment by Philius** [comment]
I promise I'm not making this up.
Now it fails either way with a different problem:
```
contact@bookworm:/v3c/notes/clang/Debug$ ~/.local/share/perfetto/prebuilts/trace_processor_shell --httpd /v3c/notes/chrome-clang.json
Illegal instruction
contact@bookworm:/v3c/notes/clang/Debug$ ~/.local/share/perfetto/prebuilts/trace_processor_shell --httpd /v3c/notes/clang/chrome-clang.json
Illegal instruction
```
Could it be the --httpd flag?
I even renamed the `trace_processor_shell` to `trace_processor_shell.bak` so the python script would re-download it.
Same result.
Instead of testing with the one you built, could you give the one available for download a try?

**Comment by Philius** [comment]
One thing did change.
I left my PC in for repair and transferred the SSD to an old Intel i7 laptop with 8G ram.

**Comment by Philius** [comment]
Any chance you can make available a `trace_processor_shell` with debugging symbols so I can do a core dump?
Maybe add a `-d` option to the python script to run `trace_processor_shell-d` or similiar.

**Comment by Philius** [comment]
I also remember that there was a recent Intel microcode update.

```
Intel(R) Core(TM) i7-2670QM CPU @ 2.20GHz (family: 0x6, model: 0x2a, stepping: 0x7)
...
Aug 01 17:50:32 bookworm kernel: MDS CPU bug present and SMT on, data leak possible. See https://www.kernel.org/doc/html/latest/admin-guide/hw-vuln/mds.html for more det>
Aug 01 17:50:32 bookworm kernel:   #5  #6  #7
Aug 01 17:50:32 bookworm kernel: smp: Brought up 1 node, 8 CPUs
...
Aug 01 17:50:32 bookworm kernel: microcode: sig=0x206a7, pf=0x10, revision=0x25
Aug 01 17:50:32 bookworm kernel: microcode: Microcode Update Driver: v2.2.
```

**Comment by LalitMaganti** [comment]
The trace processor prebuilts require the availability of AVX2 which appears to be unavailable for your CPU. We have a [check](https://cs.android.com/android/platform/superproject/main/+/main:external/perfetto/src/base/utils.cc;l=119?q=f:perfetto%20avx2) for this though, not sure why it's not firing for you.

We don't distribute debug/non-AVX binaries but it's quite easy to compile them yourself. Instructions available at https://perfetto.dev/docs/contributing/getting-started but TLDR:
```
git clone https://android.googlesource.com/platform/external/perfetto/
cd perfetto
tools/install-build-deps
tools/setup_all_configs.py
tools/gn gen out/linux_clang_debug --args="is_clang=true is_debug=true enable_perfetto_x64_cpu_opt=false"
tools/ninja -C out/linux_clang_debug
```

Then the binary will be available at `out/linux_clang_debug/trace_processor_shell`.

**Comment by Philius** [comment]
Check for a file that doesn't exist:
```
/v3c/notes/clang$ /v3c/downloads/Google/perfetto/out/linux_clang_debug/trace_processor_shell  --httpd /v3c/notes/clang/chrome-clangX.json
[964.639]     sqlite_table.cc:285 Create table statement: CREATE TABLE x( rowid BIGINT HIDDEN, quantum BIGINT HIDDEN, window_start BIGINT HIDDEN, window_dur BIGINT HIDDEN, ts BIGINT, dur BIGINT, quantum_ts BIGINT, PRIMARY KEY(rowid)) WITHOUT ROWID;
Could not read trace file (path: /v3c/notes/clang/chrome-clangX.json): Could not open trace file (path: /v3c/notes/clang/chrome-clangX.json)
```
Fails as expected.

Check for a small trace file chrome-Simple.json, size 1.4KB
(I won't show the log as there's lots of it)
Works as expected.

Check for a 32.2 GB file:
```
/v3c/notes/clang$ /v3c/downloads/Google/perfetto/out/linux_clang_debug/trace_processor_shell  --httpd /v3c/notes/clang/chrome-clang.json
[784.013]     sqlite_table.cc:285 Create table statement: CREATE TABLE x( rowid BIGINT HIDDEN, quantum BIGINT HIDDEN, window_start BIGINT HIDDEN, window_dur BIGINT HIDDEN, ts BIGINT, dur BIGINT, quantum_ts BIGINT, PRIMARY KEY(rowid)) WITHOUT ROWID;
[784.059] ding_trace_parser.cc:77 JSON trace detected
[796.433]   trace_blob_view.h:100 PERFETTO_CHECK(off + length <= length_) (errno: 0, Success)

-----BEGIN PERFETTO PRE-CRASH LOG-----
trace_uuid: 00000000-0000-0000-c895-fd2724b
[784.013]     sqlite_table.cc:285 Create table statement: CREATE TABLE x( rowid BIGINT HIDDEN, quantum BIGINT HIDDEN, window_start BIGINT HIDDEN, window_dur BIGINT HIDDEN, ts BIGINT, dur BIGINT, quantum_ts BIGINT, PRIMARY KEY(rowid)) WITHOUT ROWID;
[784.059] ding_trace_parser.cc:77 JSON trace detected
[796.433]   trace_blob_view.h:100 PERFETTO_CHECK(off + length <= length_) (errno: 0, Success)

-----END PERFETTO PRE-CRASH LOG-----

------------------ BEGINNING OF CRASH ------------------
Signal: Illegal instruction (possibly unaligned access)
Fault addr: 000056026D2813C6

Backtrace:
unrecognno debug
#00  0x00007FA80BED9FD0
no debug
#01  0x000056026D2813C6
no debug
#02  0x000056026D280D11
no debug
#03  0x000056026D258675
no debug
#04  0x000056026D25696D
no debug
#05  0x000056026D2561E3
no debug
#06  0x00007FA80BEC51CA
no debug
#07  0x00007FA80BEC5285
no debug
#08  0x000056026D2560E1
no debug
#09  0x0000000000000000
------------------ END OF CRASH ------------------
Illegal instruction
```
The `PERFETTO_CHECK()` assertion macro fails and executes an illegal instruction to end the program.
Looking into it with KDbg:
```
off = 134217728
length = 134217728
length_ = 214545941
```
I'm not sure how to proceed given that we're running differently compiled `trace_processor_shell` programs on different processors with presumably different amounts of RAM.

**Comment by primiano** [comment]
We don't support internally files bigger than 32GB for internal optimization purposes.
TBH we should improve our logging to tell that explicitly when opening the file, rather than later on via a CHECK.
In various places we use uint32_t for file size, which limits the max trace size to 32GB (which seems reasonable enough as a limit, if you ask me)

**Comment by Philius** [comment]
I'm building a simple c++ project that both generates and uses a pre-compiled headers, using ninja.
I'm tracing the ninja build using a RelWithDebugInfo build of clang to determine at what point in the call hierarchy, after reading in the precompiled header, does a compilation process eventually get around to actually reading in the source file.

I have ideas about tree view disk representations that don't require that you read the entire trace file by allowing the contents to be navigable.
mmap() a potentially multi-gigabyte trace file and only read in pages as needed. See [treedb](https://sourceforge.net/projects/treedb/).

My goal is to implement precompiled headers using treedb so that compilations running in parallel, not only use the same precompiled header, but are forks of one process that has already read it in and are about to read in the source file.

Perfetto could be modified to "drill down" to more detail as needed once an item is expanded in size to be greater than one horizontal pixel, without having to read in the whole trace.

In my use case, I'm only interested in the fifth child compilation process, so entire gigabytes of the trace file could be skipped.

**Comment by Philius** [comment]
In case I wasn't clear, implementing precompiled headers using treedb would allow different build runs to directly mmap in the precompiled header instead of de-serializing it into each individual compilation process which is how it is done now.
That memory can't be shared among processes.

If there's another way to produce a call hierarchy from a trace file, I'm all ears.

**Comment by LalitMaganti** [comment]
Thanks for working with us to figure out the root cause of this issue! I just want to make clear from our view on this bug.

While the issue is real and an important one, in our view it is quite an edge case. We're very happy that the Perfetto is useful and capable enough to visualise traces collected from sources other than from Android (our primary focus). However, in cases like this where an issue in an external usecase would require a significant investment to fix, we have to think very carefully. The core Perfetto team is quite small for the number of things we are trying to do and, thus, we choose to focus intently the projects which are most important to the success of Android.

Basically, I don't think we are going to do anything further on this bug (apart from the fix [Primiano mentioned](https://github.com/google/perfetto/issues/539#issuecomment-1662834300)). If someone wants to contribute patches towards fixing this, we'd be more than happy to accept them!

**Comment by primiano** [comment]
Hey I'm sorry but I'm closing this issue.
The main reason why I am closing this is that the issue is too unactionable and cahotic. On every comment a new problem was presented, completely independent of the previous one, without any clear repro step.

As much as we are committed to fixing bugs and improving the quality of our codebase, I'm giving up on this one. In future I'd suggest to file clear bugs, focusing on a specific problem, with clear repro steps.

Here you went from mmap errors, to new errors coming from code with random patches, to file-size related issues, to microcode updates. There's just too much. We have a limited amount of time to invest on each bug and this one has been a moving target.

Sorry but this is just too much.
Primiano
